finall

Evg33n opened the pull request:
finall

Commit:
1

finall

evpozdniakov commented on DNA/DNA/Cell.h:
```diff
@@ -0,0 +1,17 @@
+
+@interface Cell : NSObject
+   
+@property NSMutableArray *DNA;
```
В задании говорится "В классе создайте переменную DNA типа NSMutableArray". У вас вместо переменной свойство.

slebedev replied:
Ну и что, в свойстве так же создается переменная.

evpozdniakov replied:
В задании говорится: "в классе создайте переменную DNA типа NSMutableArray...".

Возможно я ошибаюсь, но мне кажется, что свойство в данном случае не является правильным решением.

Вы говорите, что в свойстве тоже создается переменная. Это так, но, во-первых, это неявная переменная, и во-вторых, с именем отличным от DNA.

slebedev replied:
Имя можно задать и такое же, опять же. И что значит "неявная переменная"?

evpozdniakov replied:
Я только что пересмотрел лекцию о свойствах.
https://u.hexlet.org/courses/4/wiki/svoistva-properties?module_item_id=7

Там говорится, что свойсвтва создают переменные, которые "запрятаны". Это я и называю "неявными" переменными, не напрямую объявленными.

Ничего не могу сказать насчет таго, каким образом с помощью свойства можно задать определенное имя переменной. Об этом нам еще не рассказывали :)

slebedev replied:
Рассказывали. С помощью @synthesize. :) 

Поясню. Если мы объявили свойство 

@property NSMutableArray *DNA;

То, если вы пользуетесь последней версией Xcode, вам писать @synthesize не обязательно, он подразумевается автоматически. Но раньше нужно было синтезировать свойство вручную. Так вот, возвращаемся к свойству. По умолчанию, пусть нам и не видно, в имплементации после задания свойства как бы присутствует строчка: 

@synthesize DNA = _DNA;

Эта запись означает, что наше свойство теперь имеет сеттеры и геттеры, а состояние переменной доступно по _DNA. Можно правую часть вообще убрать: 

@synthesize DNA;

Эта запись означает следующее:

@synthesize DNA = DNA;

Теперь внутри класса упоминание DNA означает внутреннее состояние класса - переменную NSMutableArray *DNA;

Относитесь к свойству, как к макросу, который скрывает за собой что-то определенное, что раньше делалось вручную (сеттеры, геттеры, слабые и сильные ссылки, потоки и так далее) к переменным класса.

evpozdniakov replied:
Круто, я этого не знал, спасибо за науку :)

Вы меня убедили, я не настаиваю больше на том, что нужно объявлять DNA переменной.

Мне интересно, что такое сильные и слабые ссылки? потоки? И что еще может быть объявлено в @synthesize? И поскольку это полезные сведения, то было бы хорошо создать для этого отдельную тему в дискуссиях.

Спасибо!

slebedev replied:
Уверен, в курсе еще будет про это, не беспокойтесь. :) Кстати, я имел в виду не "потоки", а "потоковость", что ли... В общем, в свойстве можно сделать переменную подготовленной для работы с несколькими потоками. 

Просматривая различные варианты, вы могли видеть следующий вариант: 

@property (nonatomic,readonly) NSMutableArray *DNA;

nonatomic - без потоков 
readonly - только для чтения (т.е. автоматически создается только геттер)

weak, strong - тип "владения" переменной (слабые и жесткие ссылки), об этом мы по-любому будем говорить, когда затронем работу памяти. 

В общем, "оставайтесь с нами", все самое интересное еще впереди! ;)

evpozdniakov commented on DNA/DNA/Cell.h:
```diff
@@ -0,0 +1,17 @@
+
+-(void) print;
+
+@end
```
В задании говорится "Создайте метод hammingDistance, который...". У вас в интерфейсе этот метод отсутствует.

evpozdniakov commented on DNA/DNA/Cell.m:
```diff
@@ -0,0 +1,35 @@
+    }
+    return self;
+}
+-(void) print {
```
Если уж вы описали данный метод в имплементации, следовало бы его добавить и в интерфейс.

evpozdniakov commented on DNA/DNA/main.m:
```diff
@@ -0,0 +1,24 @@
+        Cell *dnaTest = [[Cell alloc] init];
+       
+        [dnaTest print];
+        
```
Задание выполнено не полностью: из 5 пунктов выполнены только 2. Я бы оценил вашу работу на 4 балла из 10. Пожалуйста, сообщите, будете ли вы доделывать задание. 

Удачи!

Evg33n replied:
Задание буду еще доделывать, просто не было времени и решил отправить что было готово. Спасибо учту все описанные замечания и доделаю за выходные проект.

Commit:
2

Да задание еще буду доделовать просто тогда не было времени и отправил
что сделат

Commit:
Revert "2"

This reverts commit 287e0723563df377b8f0ffc88cfe8440872ed269.

Evg33n commented:
Задание буду еще доделывать, просто не было времени и решил отправить что было готово. Спасибо учту все описанные замечания и доделаю за выходные проект.

olegto commented:
Дайте знать как будет готово.

Commit:
*FINAL*

последний вариант больше изменять не буду

Commit:
final

готовая работа
